File: multi-threaded-santa/python/src/multi_threaded_santa/bounded_queue.py
import queue
import threading
from typing import Generic, TypeVar

T = TypeVar("T")


class BoundedQueue(Generic[T]):
    """Thread-safe bounded queue backed by queue.Queue.

    Producers block when full; consumers block when empty.
    Call complete() to signal no more items will be enqueued.
    """

    def __init__(self, capacity: int) -> None:
        self.capacity = capacity
        self._queue: queue.Queue[T] = queue.Queue(maxsize=capacity)
        self._completed = threading.Event()

    @property
    def count(self) -> int:
        return self._queue.qsize()

    def enqueue(self, item: T, timeout: float | None = None) -> bool:
        """Enqueue an item. Blocks if full. Returns False if completed."""
        if self._completed.is_set():
            return False
        try:
            self._queue.put(item, timeout=timeout)
            return True
        except queue.Full:
            return False

    def try_enqueue(self, item: T) -> bool:
        """Try to enqueue without blocking. Returns False if full or completed."""
        if self._completed.is_set():
            return False
        try:
            self._queue.put_nowait(item)
            return True
        except queue.Full:
            return False

    def dequeue(self, timeout: float = 0.1) -> tuple[bool, T | None]:
        """Dequeue an item. Blocks if empty.

        Returns (True, item) or (False, None) when completed and empty.
        Uses a polling loop so it can detect completion.
        """
        while True:
            try:
                item = self._queue.get(timeout=timeout)
                return True, item
            except queue.Empty:
                if self._completed.is_set() and self._queue.empty():
                    return False, None

    def complete(self) -> None:
        """Signal that no more items will be enqueued."""
        self._completed.set()

    @property
    def is_completed(self) -> bool:
        return self._completed.is_set()

    def items(self) -> "BoundedQueueIterator[T]":
        """Iterator that yields items until the queue is completed and empty."""
        return BoundedQueueIterator(self)


class BoundedQueueIterator(Generic[T]):
    def __init__(self, bounded_queue: BoundedQueue[T]) -> None:
        self._queue = bounded_queue

    def __iter__(self) -> "BoundedQueueIterator[T]":
        return self

    def __next__(self) -> T:
        success, item = self._queue.dequeue()
        if not success:
            raise StopIteration
        return item  # type: ignore[return-value]

File: multi-threaded-santa/csharp/src/MultiThreadedSanta/BoundedQueue.cs
using System.Threading.Channels;

namespace MultiThreadedSanta;

/// <summary>
/// Thread-safe bounded queue backed by <see cref="Channel{T}"/>.
/// Producers block when full; consumers block when empty.
/// Call <see cref="Complete"/> to signal no more items will be enqueued.
/// </summary>
public sealed class BoundedQueue<T>
{
    private readonly Channel<T> _channel;

    public int Capacity { get; }

    public BoundedQueue(int capacity)
    {
        Capacity = capacity;
        _channel = Channel.CreateBounded<T>(new BoundedChannelOptions(capacity)
        {
            FullMode = BoundedChannelFullMode.Wait,
            SingleWriter = false,
            SingleReader = false
        });
    }

    /// <summary>Returns the number of items currently in the queue.</summary>
    public int Count => _channel.Reader.Count;

    /// <summary>
    /// Enqueue an item. Blocks (asynchronously) if the queue is at capacity.
    /// Returns false if the queue has been completed.
    /// </summary>
    public async ValueTask<bool> EnqueueAsync(T item, CancellationToken ct = default)
    {
        try
        {
            await _channel.Writer.WriteAsync(item, ct);
            return true;
        }
        catch (ChannelClosedException)
        {
            return false;
        }
    }

    /// <summary>
    /// Try to enqueue an item without waiting.
    /// Returns false if the queue is full or completed.
    /// </summary>
    public bool TryEnqueue(T item)
    {
        return _channel.Writer.TryWrite(item);
    }

    /// <summary>
    /// Dequeue an item. Blocks (asynchronously) if the queue is empty.
    /// Returns (true, item) or (false, default) when the queue is completed and empty.
    /// </summary>
    public async ValueTask<(bool Success, T? Item)> DequeueAsync(CancellationToken ct = default)
    {
        try
        {
            var item = await _channel.Reader.ReadAsync(ct);
            return (true, item);
        }
        catch (ChannelClosedException)
        {
            return (false, default);
        }
    }

    /// <summary>
    /// Signal that no more items will be enqueued.
    /// Consumers will drain remaining items and then see completion.
    /// </summary>
    public void Complete()
    {
        _channel.Writer.TryComplete();
    }

    /// <summary>Async enumerable over all items until completion.</summary>
    public IAsyncEnumerable<T> ReadAllAsync(CancellationToken ct = default)
    {
        return _channel.Reader.ReadAllAsync(ct);
    }
}
